test/with_dummyserver/test_connectionpool.py

```diff
@@ -870,5 +870,58 @@ def test_redirect_after(self):
         self.assertEqual(r.status, 200)
         self.assertTrue(delta < 1)
 
+class TestFileBodiesOnRetryOrRedirect(HTTPDummyServerTestCase):
+    def setUp(self):
+        self.pool = HTTPConnectionPool(self.host, self.port, timeout=0.1)
+
+    def test_retries_put_filehandle(self):
+        """HTTP PUT retry with a file-like object should not timeout"""
+        retry = Retry(total=3, status_forcelist=[418])
+        # httplib reads in 8k chunks; use a larger content length
+        content_length = 65535
+        data = b'A' * content_length
+        uploaded_file = io.BytesIO(data)
+        headers = {'test-name': 'test_retries_put_filehandle',
+                   'Content-Length': str(content_length)}
+        resp = self.pool.urlopen('PUT', '/successful_retry',
+                                 headers=headers,
+                                 retries=retry,
+                                 body=uploaded_file,
+                                 assert_same_host=False, redirect=False)
+        self.assertEqual(resp.status, 200)
+
+    def test_redirect_put_file(self):
+        """PUT with file object should work with a redirection response"""
+        retry = Retry(total=3, status_forcelist=[418])
+        # httplib reads in 8k chunks; use a larger content length
+        content_length = 65535
+        data = b'A' * content_length
+        uploaded_file = io.BytesIO(data)
+        headers = {'test-name': 'test_redirect_put_file',
+                   'Content-Length': str(content_length)}
+        resp = self.pool.urlopen('PUT', '/echo',
+                                 headers=headers,
+                                 retries=retry,
+                                 body=uploaded_file,
+                                 assert_same_host=False, redirect=True)
```

Comment:
So this new test is good, but does it..uh, does it ever actually rewind the body?